tests/integration_test/transport_security_test.go

```diff
@@ -214,3 +214,62 @@ func TestHttpsClients(t *testing.T) {
 		}
 	}
 }
+
+func TestHSTS(t *testing.T) {
+	t.Parallel()
+	args := utils.CommonArgs()
+	args = append(args, "--ssl_server_cert_path=../env/testdata/")
+	args = append(args, "--enable_strict_transport_security")
+
+	s := env.NewTestEnv(comp.TestHttpsClients, platform.EchoSidecar)
+	defer s.TearDown()
+	s.AppendHttpRules([]*annotationspb.HttpRule{
+		{
+			Selector: "1.echo_api_endpoints_cloudesf_testing_cloud_goog.Simpleget",
+			Pattern: &annotationspb.HttpRule_Get{
+				Get: "/simpleget",
+			},
+		},
+	})
+	if err := s.Setup(args); err != nil {
+		t.Fatalf("fail to setup test env, %v", err)
+	}
+
+	testData := []struct {
+		desc           string
+		httpsVersion   int
+		certPath       string
+		wantHSTSHeader string
+		wantResp       string
+	}{
+		{
+			desc:           "Succcess for HTTP1 client with HSTS",
+			httpsVersion:   1,
+			certPath:       platform.GetFilePath(platform.ServerCert),
+			wantHSTSHeader: "max-age=31536000; includeSubdomains",
+		},
+		{
+			desc:           "Succcess for HTTP2 client with HSTS",
+			httpsVersion:   2,
+			certPath:       platform.GetFilePath(platform.ServerCert),
+			wantHSTSHeader: "max-age=31536000; includeSubdomains",
+		},
+	}
+
+	for _, tc := range testData {
+		url := fmt.Sprintf("https://localhost:%v/simpleget?key=api-key", s.Ports().ListenerPort)
```

Comment:
Could we have a test, that client trying to downgrade https to http(I don't know how that work, but as my understanding this is what HSTS is trying to prevent).  then test it, without HSTS, it is successful, it will fail such downgrade it it is enabled.

Reply:
HSTS (like same-origin policy) is a protocol that **browsers** must adhere to. I don't think we can test this in our integration tests. All Envoy is doing is adding a header in the responses, it's the browser that prevents using HTTP.

Reply:
Agree, we just need to make sure the header is returned.